BREAKING(sasl): make sasl mechanisms asynchronous

pierreca opened the pull request:
Related to #110 

**not ready/complete yet**

this change allows a user of the library to pass an already instantiated SASL mechanism that could respond to SASL challenges in an asynchronous manner.

I'm running into a race condition that I don't know how to solve though:

the `SaslClient.has_writes_pending()` method is called while the SASL mechanism is still working on the response (before it calls the callback of line 256). This is true but since no frames are enqueued, nothing is sent. Then when the callback is called, it's too late and nothing is calling either has_writes_pending() or `SaslClient.write()`. makes me think i'm either not using the right pattern for asynchronism with the library. I do not understand how/why `has_writes_pending()` is called.
FYI I can make things work by calling `self.connection.output()` on line 262 so I'm guessing whatever loop/event is used to call `has_writes_pending()` (and subsequently `write()`) fires prematurely or should loop until a `write()` has actually happened if `has_write_pending()` returns true?



PS: still lacking tests - want to validate the pattern and make things work first.

[edit] clarification/additional info on the race condition.

grs commented:
@pierreca I think calling connection.output() is the right thing at present when some asynchornous callback returns meaning that the sasl exchange can proceed.

grs commented on lib/sasl.js:
```diff
@@ -228,25 +228,39 @@ SaslClient.prototype.on_sasl_mechanisms = function (frame) {
         var f = this.mechanisms[mech];
         if (f) {
-            this.mechanism = f();
+            this.mechanism = typeof f === 'function' ? f() : f;
```
Can you explain this a little? (Not saying it is wrong, just want to understand better).

pierreca replied:
sure: in the current state of things, if I create my own custom SASL mechanism class, I have to let the client instantiate it with no parameters (it calls `f()` to instantiate it). In my case this design makes my life more difficult because I want to be able, when I instantiate the SASL mechanism, to give it a couple of parameters and a reference to the object that interfaces with the TPM chip.

I made this change so that if I give rhea an object that is already instantiated, it does not try to create it, it just takes it "as is". Does that make sense?

Not saying it's the right way to do it. this was the quick way :)

grs replied:
makes sense, thanks!

pierreca commented:
@grs Added calls to `connection.output()` per your first comment. Obviously i'll get test coverage where it needs to be, before I do that though, i'd like to hear your thoughts on the overall pattern. rhea is generally not callback-oriented so it's a little weird to use them only here, and I guess we could switch to events: If you have a pattern in mind that you would like better, please do let me know :)

pierreca commented:
@grs any other thoughts on this PR and the suggested callback pattern or should I start banging out tests?

grs commented:
@pierreca it looks good to me

pierreca commented:
@grs 
It's now rebased, squashed and ready to be merged. We'll be eagerly waiting for a new version of the package containing those changes :) (which should probably be 0.3.0 since it's technically a breaking change? your call).

amarzavery commented on lib/sasl.js:
```diff
@@ -71,13 +71,13 @@ var PlainClient = function(username, password) {
 };
 
-PlainClient.prototype.start = function() {
+PlainClient.prototype.start = function(callback) {
```
@pierreca - Can you also update the type definitions? In this way things are consistent.
[PlainClient.start](https://github.com/amqp/rhea/blob/master/typings/sasl.d.ts#L55) should now take a callback.

amarzavery commented on lib/sasl.js:
```diff
@@ -94,11 +94,11 @@ var AnonymousClient = function(name) {
 };
 
-AnonymousClient.prototype.start = function() {
+AnonymousClient.prototype.start = function(callback) {
```
same as above.

amarzavery commented on lib/sasl.js:
```diff
@@ -114,20 +114,20 @@ var ExternalClient = function() {
 };
 
-XOAuth2Client.prototype.start = function() {
+XOAuth2Client.prototype.start = function(callback) {
```
update type definition

amarzavery commented on lib/sasl.js:
```diff
@@ -114,20 +114,20 @@ var ExternalClient = function() {
-ExternalClient.prototype.step = function() {
-    return '';
+ExternalClient.prototype.step = function(callback) {
+    callback(undefined, '');
```
update type definition

amarzavery commented on lib/sasl.js:
```diff
@@ -114,20 +114,20 @@ var ExternalClient = function() {
 
-ExternalClient.prototype.start = function() {
-    return '';
+ExternalClient.prototype.start = function(callback) {
```
update type definition

pierreca commented:
@amarzavery good catch. I updated the type definitions. LMK if you'd like to see something else - if it's all good I'll rebase again and let @grs merge/release
